lighthouse-core/audits/treemap-data.js

```diff
@@ -0,0 +1,309 @@
+        }
+      });
+    }
+
+    const rootNode = newNode(sourceRoot);
+
+    // For every source file, apply the data to all components
+    // of the source path, creating nodes as necessary.
+    for (const [source, data] of Object.entries(sourcesData)) {
+      addNode(source || `<unmapped>`, data, rootNode);
+
+      // Apply the data to the rootNode.
+      rootNode.resourceBytes += data.resourceBytes;
+      if (data.unusedBytes) rootNode.unusedBytes = (rootNode.unusedBytes || 0) + data.unusedBytes;
+    }
+
+    /**
+     * Collapse nodes that have only one child.
+     * @param {Node} node
+     */
+    function collapse(node) {
+      while (node.children && node.children.length === 1) {
+        node.name += '/' + node.children[0].name;
+        node.children = node.children[0].children;
+      }
+
+      if (node.children) {
+        for (const child of node.children) {
+          collapse(child);
+        }
+      }
+    }
+    collapse(rootNode);
+
+    // TODO(cjamcl): Should this structure be flattened for space savings?
+    // Like DOM Snapshot.
+    // Less JSON (no super nested children, and no repeated property names).
+
+    return rootNode;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<RootNode[]>}
+   */
+  static async makeJavaScriptRootNodes(artifacts, context) {
+    /** @type {RootNode[]} */
+    const rootNodes = [];
+
+    const bundles = await JsBundles.request(artifacts, context);
+    const duplication = await ModuleDuplication.request(artifacts, context);
+
+    /** @type {Array<{src: string, length: number, unusedJavascriptSummary?: import('../computed/unused-javascript-summary.js').Summary}>} */
+    const scriptData = [];
+    const inlineScriptData = {
+      src: artifacts.URL.finalUrl,
+      length: 0,
+    };
+    for (const scriptElement of artifacts.ScriptElements) {
+      // Normalize ScriptElements so that inline scripts show up as a single entity.
+      if (!scriptElement.src) {
+        inlineScriptData.length += (scriptElement.content || '').length;
+        continue;
+      }
+
+      const url = scriptElement.src;
+      const bundle = bundles.find(bundle => url === bundle.script.src);
+      const scriptCoverages = artifacts.JsUsage[url];
+      if (!bundle || !scriptCoverages) continue;
+
+      scriptData.push({
+        src: scriptElement.src,
+        length: (scriptElement.content || '').length,
+        unusedJavascriptSummary:
+          await UnusedJavaScriptSummary.request({url, scriptCoverages, bundle}, context),
+      });
+    }
+    if (inlineScriptData.length) scriptData.unshift(inlineScriptData);
+
+    for (const {src, length, unusedJavascriptSummary} of scriptData) {
+      const bundle = bundles.find(bundle => bundle.script.src === src);
+      const name = src;
+
+      let node;
+      if (bundle && unusedJavascriptSummary && unusedJavascriptSummary.sourcesWastedBytes) {
```

Comment:
style nit: totally my opinion, but reversing the order of these conditionals (and possibly early push and `continue`s?) might help with readability here too. e.g.

```js
if (!unusedJavascriptSummary) {
  // Don't even have coverage, so only provide the byte length.
  // (is this even possible if there's a guaranteed `scriptCoverages`?)
  // ...
} else if (!unusedJavascriptSummary.sourcesWastedBytes)  {
  // Without sources, can't delve inside, so give bundle-level stats.
  // ...
} else {
  // Full source-map-based data.
}
```